src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs

```diff
@@ -279,47 +279,60 @@ public override void Dispose()
-                    {
-                        messageType = WebSocketMessageType.Binary;
-                    }
-                    else
+                    uint dataAvailable;
+                    while ((dataAvailable = reader.UnconsumedBufferLength) > 0)
                     {
-                        messageType = WebSocketMessageType.Text;
-                    }
+                        ArraySegment<byte> buffer;
+                        try
+                        {
+                            buffer = _receiveAsyncBufferTcs.Task.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException) // Caused by Abort call on WebSocket
+                        {
+                            return;
+                        }
 
-                    bool endOfMessage = false;
-                    uint readCount = Math.Min(dataAvailable, (uint) buffer.Count);
-                    var dataBuffer = reader.ReadBuffer(readCount);
-                    // Safe to cast readCount to int as the maximum value that readCount can be is buffer.Count.
-                    dataBuffer.CopyTo(0, buffer.Array, buffer.Offset, (int) readCount);
-                    if (dataAvailable == readCount)
-                    {
-                        endOfMessage = true;
-                    }
+                        _receiveAsyncBufferTcs = new TaskCompletionSource<ArraySegment<byte>>();
+                        WebSocketMessageType messageType;
+                        if (args.MessageType == SocketMessageType.Binary)
+                        {
+                            messageType = WebSocketMessageType.Binary;
+                        }
+                        else
+                        {
+                            messageType = WebSocketMessageType.Text;
+                        }
 
-                    WebSocketReceiveResult recvResult = new WebSocketReceiveResult((int) readCount, messageType,
-                        endOfMessage);
-                    _webSocketReceiveResultTcs.TrySetResult(recvResult);
+                        bool endOfMessage = false;
+                        uint readCount = Math.Min(dataAvailable, (uint) buffer.Count);
+                        var dataBuffer = reader.ReadBuffer(readCount);
+                        // Safe to cast readCount to int as the maximum value that readCount can be is buffer.Count.
+                        dataBuffer.CopyTo(0, buffer.Array, buffer.Offset, (int) readCount);
+                        if (dataAvailable == readCount)
+                        {
+                            endOfMessage = true;
+                        }
+
+                        WebSocketReceiveResult recvResult = new WebSocketReceiveResult((int) readCount, messageType,
+                            endOfMessage);
+                        _webSocketReceiveResultTcs.TrySetResult(recvResult);
+                    }
                 }
             }
+            catch (Exception)
```

Comment:
We usually do not catch all exceptions in .NET code because there are cases where we want "unexpected" exceptions to crash the process.  So, perhaps there is a more specific type of exception to catch here?

Also, we probably should save this exception in the catch block and then use it as an inner exception when you finally throw the WebSocketException later on. This would greatly help with diagnosing these failures.